File: Day51/Part1/crystal_labyrinth.py
from collections import deque

def solve_crystal_labyrinth(grid, K):
    N = len(grid)
    M = len(grid[0])
    
    start_pos = None
    end_pos = None
    teleports = []
    crystals = [] # (r, c, value)
    
    for r in range(N):
        for c in range(M):
            cell = grid[r][c]
            if cell == 'S':
                start_pos = (r, c)
            elif cell == 'E':
                end_pos = (r, c)
            elif cell == 'T':
                teleports.append((r, c))
            elif '1' <= cell <= '9':
                crystals.append((r, c, int(cell)))
                
    num_crystals = len(crystals)
    
    # Map (r, c) to crystal index
    crystal_map = {}
    for i, (r, c, val) in enumerate(crystals):
        crystal_map[(r, c)] = i
        
    # BFS State: (r, c, mask) -> min_steps
    # Initialize with infinity
    # Using a dictionary or 3D array? 3D array is faster.
    # dist[r][c][mask]
    # But mask can be up to 2^9 = 512.
    
    # Optimization: If K is small, maybe we don't visit all states?
    # But K can be large.
    
    queue = deque()
    
    # Initial state
    start_mask = 0
    if start_pos in crystal_map:
        start_mask |= (1 << crystal_map[start_pos])
        
    queue.append((start_pos[0], start_pos[1], start_mask, 0))
    
    visited = set() # (r, c, mask)
    visited.add((start_pos[0], start_pos[1], start_mask))
    
    max_value = -1
    
    # Directions
    dirs = [(0, 1), (0, -1), (1, 0), (-1, 0)]
    
    while queue:
        r, c, mask, steps = queue.popleft()
        
        # Check if reached E
        if (r, c) == end_pos:
            # Calculate value
            current_val = 0
            for i in range(num_crystals):
                if (mask >> i) & 1:
                    current_val += crystals[i][2]
            max_value = max(max_value, current_val)
            
        if steps >= K:
            continue
            
        # Generate next moves
        next_positions = []
        
        # 1. Normal moves
        for dr, dc in dirs:
            nr, nc = r + dr, c + dc
            if 0 <= nr < N and 0 <= nc < M and grid[nr][nc] != '#':
                next_positions.append((nr, nc))
                
        # 2. Teleport
        # If current is T, we can instantly be at any other T
        # But "Teleport tidak menghitung langkah".
        # This means if we are at T, we are effectively at ALL Ts simultaneously for 0 cost.
        # So from a T, we can move to neighbors of ANY T.
        # Or simpler: If we step ONTO a T, we add all Ts to the queue with same steps?
        # Wait, if we step onto T at step S, we are at all Ts at step S.
        # So we should expand from all Ts.
        # To avoid infinite loops, we handle this carefully.
        # Better: If grid[r][c] == 'T', we add all other Ts to queue with SAME steps and SAME mask?
        # Yes, but mark them visited to avoid cycles.
        
        if grid[r][c] == 'T':
            for tr, tc in teleports:
                if (tr, tc) == (r, c): continue
                if (tr, tc, mask) not in visited:
                    visited.add((tr, tc, mask))
                    queue.appendleft((tr, tc, mask, steps)) # 0 cost, push front
                    
        for nr, nc in next_positions:
            new_mask = mask
            if (nr, nc) in crystal_map:
                new_mask |= (1 << crystal_map[(nr, nc)])
                
            if (nr, nc, new_mask) not in visited:
                visited.add((nr, nc, new_mask))
                queue.append((nr, nc, new_mask, steps + 1))
                
    return max_value if max_value != -1 else 0

if __name__ == "__main__":
    grid = [
        ['S', '.', '1', '#', 'E'],
        ['.', '#', '.', '#', '.'],
        ['T', '.', '2', '.', 'T']
    ]
    K = 10
    # Path: S(0,0) -> (0,1) -> (0,2)[Get 1] -> (1,2) -> (2,2)[Get 2] -> (2,1) -> (2,0)[T] -> Teleport to (2,4)[T] -> (1,4) -> (0,4)[E]
    # Steps: 1, 2, 3, 4, 5, 6 (at T), 6 (at T2), 7, 8. Total 8 steps.
    # Value: 1 + 2 = 3.
    
    print(f"Max Value: {solve_crystal_labyrinth(grid, K)}")

File: Day51/Part1/crystal_labyrinth.cpp
#include <iostream>
#include <vector>
#include <string>
#include <queue>
#include <tuple>
#include <map>
#include <set>

using namespace std;

struct State {
    int r, c, mask, steps;
};

int solve_crystal_labyrinth(const vector<string>& grid, int K) {
    int N = grid.size();
    int M = grid[0].size();
    
    int start_r, start_c, end_r, end_c;
    vector<pair<int, int>> teleports;
    vector<tuple<int, int, int>> crystals;
    
    for (int r = 0; r < N; ++r) {
        for (int c = 0; c < M; ++c) {
            char cell = grid[r][c];
            if (cell == 'S') {
                start_r = r; start_c = c;
            } else if (cell == 'E') {
                end_r = r; end_c = c;
            } else if (cell == 'T') {
                teleports.push_back({r, c});
            } else if (cell >= '1' && cell <= '9') {
                crystals.push_back({r, c, cell - '0'});
            }
        }
    }
    
    int num_crystals = crystals.size();
    map<pair<int, int>, int> crystal_map;
    for (int i = 0; i < num_crystals; ++i) {
        crystal_map[{get<0>(crystals[i]), get<1>(crystals[i])}] = i;
    }
    
    // BFS
    deque<State> q;
    
    int start_mask = 0;
    if (crystal_map.count({start_r, start_c})) {
        start_mask |= (1 << crystal_map[{start_r, start_c}]);
    }
    
    q.push_back({start_r, start_c, start_mask, 0});
    
    // Visited: visited[r][c][mask]
    // Use a set or flattened vector?
    // N, M <= 50, Mask <= 512. 50*50*512 = 1.28e6. Vector is fine.
    // vector<vector<vector<bool>>> visited(N, vector<vector<bool>>(M, vector<bool>(1 << num_crystals, false)));
    // But num_crystals is dynamic. Use set for safety or flat vector.
    // Let's use set for simplicity in C++ unless TLE.
    set<tuple<int, int, int>> visited;
    visited.insert({start_r, start_c, start_mask});
    
    int max_value = -1;
    int dr[] = {0, 0, 1, -1};
    int dc[] = {1, -1, 0, 0};
    
    while (!q.empty()) {
        State curr = q.front();
        q.pop_front();
        
        if (curr.r == end_r && curr.c == end_c) {
            int current_val = 0;
            for (int i = 0; i < num_crystals; ++i) {
                if ((curr.mask >> i) & 1) {
                    current_val += get<2>(crystals[i]);
                }
            }
            if (current_val > max_value) max_value = current_val;
        }
        
        if (curr.steps >= K) continue;
        
        // Teleport logic
        if (grid[curr.r][curr.c] == 'T') {
            for (auto t : teleports) {
                if (t.first == curr.r && t.second == curr.c) continue;
                if (visited.find({t.first, t.second, curr.mask}) == visited.end()) {
                    visited.insert({t.first, t.second, curr.mask});
                    q.push_front({t.first, t.second, curr.mask, curr.steps});
                }
            }
        }
        
        for (int i = 0; i < 4; ++i) {
            int nr = curr.r + dr[i];
            int nc = curr.c + dc[i];
            
            if (nr >= 0 && nr < N && nc >= 0 && nc < M && grid[nr][nc] != '#') {
                int new_mask = curr.mask;
                if (crystal_map.count({nr, nc})) {
                    new_mask |= (1 << crystal_map[{nr, nc}]);
                }
                
                if (visited.find({nr, nc, new_mask}) == visited.end()) {
                    visited.insert({nr, nc, new_mask});
                    q.push_back({nr, nc, new_mask, curr.steps + 1});
                }
            }
        }
    }
    
    return max_value == -1 ? 0 : max_value;
}

int main() {
    vector<string> grid = {
        "S.1#E",
        ".#.#.",
        "T.2.T"
    };
    int K = 10;
    cout << "Max Value: " << solve_crystal_labyrinth(grid, K) << endl;
    return 0;
}
